src/parser/parser.cc

```diff
@@ -1527,29 +1699,50 @@ class Parser {
 
   // Parses a user defined ADT or type variable.
   Type ParseNonPrimitiveType(const Token& tok) {
-    auto name = tok.ToString();
-    Type head_type;
-    auto global_type = type_names.Get(name);
+    return WithSpan<Type>([&]() {
+      auto name = tok.ToString();
+      Type head_type = LookupTypeVar(tok);
 
-    if (!global_type) {
-      head_type = LookupTypeVar(tok);
-    } else {
-      head_type = global_type.value();
-    }
+      if (!head_type.defined()) {
+        // head_type = type_names.Get(name);
+        head_type = AddOrGet(&type_names, name, TypeKind::kAdtHandle);
+      }
 
-    CHECK(head_type.defined()) << "internal error: head type must be defined";
+      // if (!var.defined()) {
```

Comment:
We now lazily resolve names in order to defer handling, we will merge checking this into the well-formed check in follow up work to clean up the handling of this.